### PR TITLE
CA-228756: Provide more error details for precheck and apply.

### PR DESCRIPTION
**Pull request by huizh**
Catch the unexpected failure and return detailed error message.

Signed-off-by: Hui Zhang <hui.zhang@citrix.com>

### CODE REVIEW

File: scripts/extensions/pool_update.precheck [hunk 1]
```diff
@@ -140,7 +141,7 @@ def execute_precheck(session, control_package, yum_conf_file, update_precheck_fi
             for msg in livepatch_messages.keys():
                 if msg in precheck_output:
                     return livepatch_messages[msg]
-            raise PrecheckFailure()
+            raise PrecheckFailure('Unknown precheck result: %s' % precheck_output)
```
Comment (huizh) [inline scripts/extensions/pool_update.precheck]: Thanks @mseri for the comment.
In fact I just re-test and find `re.sub('\n +', ' ', output, flags=re.DOTALL)` is redundant and it doesn't help the format of the output of CLI. Without it, the output is even cleaner.
So I will remove them.

Also with discussion with @srowe, to return `Unknown precheck result: ...` is inappropriate, we should check if the precheck script runs successfully, if no raise error, if yes but no matching livepatch_message, return `ok`.

I will update the commit.

File: scripts/extensions/pool_update.precheck [hunk 2]
```diff
@@ -125,22 +124,26 @@ def execute_precheck(session, control_package, yum_conf_file, update_precheck_fi
...
-                    raise PrecheckFailure(precheckfailure_msg)
+                    raise PrecheckFailure(errmsg)
             else:
-                raise PrecheckFailure(precheckfailure_msg)
+                raise PrecheckFailure(errmsg)
         else:
-            raise PrecheckFailure()
+            raise PrecheckFailure(output)
     else:
         if not os.path.isfile(update_precheck_file):
```
Comment (srowe) [inline scripts/extensions/pool_update.precheck]: If invert this test to `if os.path.isfile(...)` and remove the `return 'ok'` ...

File: scripts/extensions/pool_update.precheck [hunk 3]
```diff
@@ -125,22 +124,26 @@ def execute_precheck(session, control_package, yum_conf_file, update_precheck_fi
...
-                    return livepatch_messages[msg]
-            raise PrecheckFailure()
+            if pp.returncode != 0:
+                raise PrecheckFailure(precheck_output)
+            else:
+                if '\n' in precheck_output:
+                    msg = precheck_output.split()[0]
+                    if msg in livepatch_messages.keys():
+                        return livepatch_messages[msg]
+                return 'ok'
```
Comment (srowe) [inline scripts/extensions/pool_update.precheck]: ... and remove one level of indent here the code is clearer.

**Comment by huizh** [comment]
Thanks @srowe and @mseri.
I've updated the commit to make the code clearer per @srowe's suggestion.

File: scripts/extensions/pool_update.precheck [hunk 4]
```diff
@@ -125,22 +124,24 @@ def execute_precheck(session, control_package, yum_conf_file, update_precheck_fi
...
                 else:
-                    raise PrecheckFailure(precheckfailure_msg)
+                    raise PrecheckFailure(errmsg)
             else:
-                raise PrecheckFailure(precheckfailure_msg)
+                raise PrecheckFailure(errmsg)
         else:
-            raise PrecheckFailure()
+            raise PrecheckFailure(output)
     else:
```
Comment (srowe) [inline scripts/extensions/pool_update.precheck]: I think this `else:` is redundant: all code paths of the `if p.returncode != 0:` result in an exception being thrown.

Can you remove it and pull the indentation of the rest of the function in one level.

Comment (huizh) [inline scripts/extensions/pool_update.precheck]: Thanks @srowe.
Just addressed it. Please could you take a look?

**Comment by srowe** [comment]
Looks good to me

**Pull request merged**